mem/idealmemcontroller/comp.go

```diff
@@ -63,8 +79,163 @@ func (c *Comp) Handle(e sim.Event) error {
 	return nil
 }
 
-// Tick updates ideal memory controller state.
 func (c *Comp) Tick() bool {
+	madeProgress := false
+
+	for i := 0; i < c.width; i++ {
+		madeProgress = c.processControlSignals() || madeProgress
+		if c.enable {
+			if c.reset {
+				madeProgress = c.updateCtrls(i) || madeProgress
+			} else {
+				madeProgress = c.updateMemCtrl() || madeProgress
+			}
+		}
+	}
+	return madeProgress
+}
+
+func (c *Comp) updateCtrls(i int) bool {
+	madeProgress := false
+	if c.reset {
+		madeProgress = c.handleReset() || madeProgress
+	} else if c.pause {
+		madeProgress = c.handlePauseProcess() || madeProgress
+	} else if c.drain {
+		madeProgress = c.handleDrain(i) || madeProgress
+	}
+	return madeProgress
+}
+
+func (c *Comp) processControlSignals() bool {
+	msg := c.CtrlPort.RetrieveIncoming()
+	if msg == nil {
+		return false
+	}
+
+	tracing.TraceReqReceive(msg, c)
+
+	switch msgType := msg.(type) {
+	case *mem.ControlMsg:
+		c.handelControlMsg(msgType)
+	}
+
+	return true
+}
+
+func (c *Comp) handelControlMsg(msg *mem.ControlMsg) {
+	if c.enable != msg.Enable {
+		c.currentEnableRsp = msg
+		c.enable = msg.Enable
+	}
+
+	if c.pause != msg.Pause {
```

Comment:
Is pause really different from `enable=false`

Reply:
I think the pause is not same as the enable=false, as shown in tick func. the enable controls the reset signal and pause signal. If we set enable = false, we cannot set reset or pause. However, if we set pause = true, we still can assign reset = true and reset the idealmemorycontroller

Reply:
Another question is what is different between reset and restart？I think we should use the same name with the same function in different components.

Reply:
Reset will erase all the internally stored data for the ideal memory controller. For idealmemorycontroller, I guess it will not be used often. But for cache and TLB, it should be used.

Reply:
Yes, I think so.